pkg/util/cache/intern.go

```diff
@@ -0,0 +1,359 @@
+		i.fileBacking.finalize()
+		i.cache = newLruStringCache(0, false)
+		i.maxStringCount = 0
+	}
+	i.refcount -= n
+}
+
+func (i *stringInterner) retain() {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	if i.refcount < 1 {
+		log.Error("Dead interner being re-retained!")
+	}
+	i.refcount++
+}
+
+// KeyedInterner has an origin-keyed set of interners.
+type KeyedInterner struct {
+	interners      *lru.Cache
+	maxQuota       int
+	closeOnRelease bool
+	tmpPath        string
+	lastReport     time.Time
+	minFileSize    int64
+	maxPerInterner int
+	lock           sync.Mutex
+}
+
+// NewKeyedStringInterner creates a Keyed String Interner with a max per-origin quota of maxQuota
+func NewKeyedStringInterner(cfg cconfig.Component) *KeyedInterner {
+	stringInternerCacheSize := cfg.GetInt("dogstatsd_string_interner_size")
+	enableMMap := cfg.GetBool("dogstatsd_string_interner_mmap_enable")
+
+	if enableMMap {
+		closeOnRelease := !cfg.GetBool("dogstatsd_string_interner_mmap_preserve")
+		tempPath := cfg.GetString("dogstatsd_string_interner_tmpdir")
+		minSizeKb := cfg.GetInt("dogstatsd_string_interner_mmap_minsizekb")
+		maxStringsPerInterner := cfg.GetInt("dogstatsd_string_interner_origin_max_strings")
+		return NewKeyedStringInternerVals(stringInternerCacheSize, closeOnRelease, tempPath, minSizeKb, maxStringsPerInterner)
+	}
+	return NewKeyedStringInternerMemOnly(stringInternerCacheSize)
+}
+
+// NewKeyedStringInternerVals takes args explicitly for initialization
+func NewKeyedStringInternerVals(stringInternerCacheSize int, closeOnRelease bool, tempPath string, minFileKb, maxStringsPerInterner int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        tempPath,
+		minFileSize:    int64(minFileKb * 1024),
+		maxPerInterner: maxStringsPerInterner,
+		closeOnRelease: closeOnRelease,
+	}
+}
+
+// NewKeyedStringInternerMemOnly is a memory-only cache with no disk needs.
+func NewKeyedStringInternerMemOnly(stringInternerCacheSize int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        "",
+		maxPerInterner: initialInternerSize,
+		closeOnRelease: false,
+	}
+}
+
+// NewKeyedStringInternerForTest is a memory-only cache with a small default size.  Useful for
+// most tests.
+func NewKeyedStringInternerForTest() *KeyedInterner {
+	return NewKeyedStringInternerMemOnly(512)
+}
+
+var sGlobalQueryCount = atomic.NewInt64(0)
+var sFailedInternalCount = atomic.NewInt64(0)
```

Comment:
`s`?

Reply:
They're static, I wasn't sure how else to denote that.  I'm completely, completely open to a better naming scheme.

Reply:
Oh I get it. Yeah, maybe a comment? Perfectly explicable once  it's explained.